src/Tokenizer/Analyzer/Analysis/MethodAnalysis.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Tokenizer\Analyzer\Analysis;
+
+/**
+ * @internal
+ */
+final class MethodAnalysis
+{
+    private ?int $visibility;
+    private bool $isStatic;
+    private bool $isAbstract;
+    private bool $isFinal;
+
+    public function __construct(?int $visibility, bool $isStatic, bool $isAbstract, bool $isFinal)
+    {
+        $this->visibility = $visibility;
+        $this->isStatic = $isStatic;
+        $this->isAbstract = $isAbstract;
+        $this->isFinal = $isFinal;
+    }
+
+    public function hasVisibility(): bool
+    {
+        return null !== $this->visibility;
+    }
+
+    public function isPublic(): bool
+    {
+        return T_PUBLIC === $this->visibility;
```

Comment:
In fact, it should be:
```suggestion
        return T_PUBLIC === $this->visibility || null === $this->visibility;
```
because without explicit visibility it's public.

Reply:
I don't like `is*VisibilityDeclared()` convention after changes, because for `protected` and `private` the only way to achieve these levels is through explicit declaration. That's why `isProtected()` and `isPrivate()` is enough, because it means that visibility was set explicitly. In terms of `public` visibility, this is why I [suggested](#discussion_r1404241410) `hasExplicitVisibility()`, because it would be possible to distinguish explicit and implicit `isPublic()` visibility.

Reply:
When coding a method in the class there are 4 options:
- do not declare visibility
- declare public
- declare protected
- declare private

We do not really care how it works, but how it is written, we have it well described in [`visibility_required`](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/v3.40.0/doc/rules/class_notation/visibility_required.rst) rule:

> Visibility MUST be declared on all properties and methods;

which implies that there are situations where visibility is not declared.

I want the names to be as close to the real meaning of what it does - e.g. to check if public visibility is declared, thus `isPublicVisibilityDeclared`, not `hasExplicitVisibility`, which requires thinking about when visibility is explicit and when implicit.

Reply:
I understand, still I think that in terms of protected and private visibility the names are redundant - `is Protected()` and `isPrivate()` is enough, because if they return true it means that visibility _is_ explicitly declared.

Looking from use cases perspective, imagine that:
- you don't want to rename public methods, so checking `isPublic()` is enough (no matter if implicit or explicit)
- you want to ensure visibility is defined, you just check `hasExplicitVisibility()` and then make decision based on that info (visibility level does not matter, you just need info if keyword is there)

That's why I suggested those changes, because I think atomic methods are better and for public visibility, if you really need distinguish implicit/explicit declaration you can just combine 2 methods (`isPublic() && !hasExplicitVisibility()`).